### DISCUSSION TITLE
"Optimize imports" inspection does not seem to work correctly when using the `qodana-jvm` linter

### DISCUSSION THREAD

**Discussion by pvlysidis** [JetBrains/Qodana#121] [Q&A] [answered]
I was testing out the later `qodana-jvm:2022.2-eap` linter together with the `--changes` flag and a custom profile by simply injecting an unused import. Once the run was done, the results were empty which was not what I expected.

Some simple replication steps using the [fastjson](https://github.com/alibaba/fastjson) project:
- Clone the repo from https://github.com/alibaba/fastjson
- Change any class in main source tree(e.g. `com.alibaba.fastjson.JSONPath`) and add an unused import. 
- Copy the default IDEA Inspection profile without changing anything and use it in `qodana.yaml`. Here are the ones I used:
 ###### `Project_Default_copy.xml `
```
<component name="InspectionProjectProfileManager">
  <profile version="1.0">
    <option name="myName" value="Project Default copy" />
  </profile>
</component>
```
  ######  `qodana.yaml`
```
version: "1.0"
profile:
  name: "Project Default copy"
```
- Execute the following command(adjust accordingly):
```
docker run -u $(id -u):$(id -g) --rm -it -p 8080:8080 \
  -v ~/sourcegit/fastjson/:/data/project/ \
  -v ~/qodana/qodana-output/fastjson_changes_only/:/data/results/ \
  -v ~/qodana/qodana-cache/:/data/cache/ \
  -v $HOME/.m2/:/root/.m2/ \
  -e _JAVA_OPTIONS=-Xmx6g \
  jetbrains/qodana-jvm:latest \
  --property=idea.required.plugins.id=Git4Idea \
  --changes \
  --show-report
```
The results of the run are empty.

Here is an example log from one of my tests: 
[idea.zip](https://github.com/JetBrains/Qodana/files/9396387/idea.zip). There are 2 `NoClassDefFoundError`s there that seem to be the cause.

Additionally, testing **without** the `-changes` flag, I see that the "Optimize imports" inspection contains results for only the test code, while if I run the inspection on the whole project via IDEA it reports 569 warnings. The "Checks" tab also does not seem to load/work properly.

Finally, doing the exact same tests, both with and without the `--changes` flag, using the `qodana-jvm-community` linter(using versions 2022.2 and 2021.3) worked just fine. Results are properly shown and the "Checks" tab works correctly as well.

This looks like a bug but I could be missing something here. Am I mis-configuring something? Let me know if you need any more information, logs, etc.

Thanks in advance!

**Comment by rillig** [comment, marked as answer]
Thanks for this bug report. Your expectations are all reasonable, and your approach should have worked.

The exceptions in `idea.log` indeed look unhealthy, we'll look into fixing them.

We already know that the `--changes` option is not 100% reliable, in an internal test configuration it sometimes fails randomly by not reporting some problems. I'm working to fix that. Due to that, your "worked just fine" _might_ actually be "seemed to work fine". We will take care of that as well.

When you say "does not seem to load/work properly", can you be a bit more specific? What exactly happens, and what do you expect instead?

**Comment by pvlysidis** [comment]
Thanks for looking into this.

> When you say "does not seem to load/work properly", can you be a bit more specific? What exactly happens, and what do you expect instead?

The number of checks and categories is not visible under the Checks tab title, instead a placeholder(like in-progress/loading) indicator is shown. Within the tab, the Applied Checks are not shown. Here is a screen shot to make it more clear:
![Screenshot from 2022-08-23 11-56-19](https://user-images.githubusercontent.com/73835071/186117049-c96798c5-16c5-4f89-9645-5ca420270c09.png)

And another screen shot when using the community linter for comparison:
![Screenshot from 2022-08-23 12-00-46](https://user-images.githubusercontent.com/73835071/186117379-01496345-87e5-4e91-8b53-8be298240150.png)

**Comment by sschuberth** [comment]
> by simply injecting an unused import.

I was doing the same in https://github.com/oss-review-toolkit/ort/pull/5805, and while Qodana correctly determines [`Yarn.kt` to be modified](https://github.com/oss-review-toolkit/ort/actions/runs/3060722462/jobs/4939583064#step:3:111), the unused import is not detected. Is that a matter of the 'qodana.starter' profile being used by default?

**Reply by sschuberth** [reply]
> Is that a matter of the 'qodana.starter' profile being used by default?

Turns out it is. The 'qodana.starter' peforms only [very limited checks](https://www.jetbrains.com/help/qodana/qodana-yaml.html#Default+profiles). Switching to the 'qodana.recommended' profile causes the unused import to be properly reported.
